src/compiler/args.rs

```diff
@@ -0,0 +1,820 @@
+use std;
+use std::cmp::Ordering;
+use std::ffi::OsString;
+use std::path::PathBuf;
+
+pub type Delimiter = Option<u8>;
+
+/// Representation of a parsed argument
+#[derive(PartialEq, Clone, Debug)]
+pub enum Argument {
+    /// Unknown non-flag argument ; e.g. "foo"
+    Raw(OsString),
+    /// Unknown flag argument ; e.g. "-foo"
+    UnknownFlag(OsString),
+    /// Known flag argument ; e.g. "-bar"
+    Flag(&'static str),
+    /// Known argument with a value ; e.g. "-qux bar", where the way the
+    /// value is passed is described by the ArgDisposition<Delimiter> type.
+    WithValue(&'static str, ArgumentValue, ArgDisposition<Delimiter>),
+}
+
+/// How a value is passed to an argument with a value.
+#[derive(PartialEq, Clone, Debug)]
+pub enum ArgDisposition<T> {
+    /// As "-arg value"
+    Separated,
+    /// As "-arg value", but "-arg<delimiter>value" would be valid too
+    CanBeConcatenated(T),
+    /// As "-arg<delimiter>value", but "-arg value" would be valid too
+    CanBeSeparated(T),
+    /// As "-arg<delimiter>value"
+    Concatenated(T),
+}
+
+pub use self::ArgDisposition::*;
```

Comment:
Is it? That's technically how you get Ok(), Err(), Some() and None for Result and Option...